execute/dispatcher.go

```diff
@@ -136,8 +145,38 @@ func (d *poolDispatcher) run(ctx context.Context) {
 		case <-d.closing:
 			// We are done, nothing left to do.
 			return
-		case fn := <-d.work:
-			fn(ctx, d.throughput)
+		case <-d.ready:
+			// Work is in the queue.
+			d.doWork(ctx)
+		}
+	}
+}
+
+// doWork will continue pulling work from the work queue
+// and running the scheduled functions until the context is canceled,
+// the dispatcher is closed, or there is no more work in the queue.
+func (d *poolDispatcher) doWork(ctx context.Context) {
+	for {
+		var fn ScheduleFunc
+		d.workMu.Lock()
+		if next := d.work.Next(); next != nil {
+			fn = next.(ScheduleFunc)
+		}
+		d.workMu.Unlock()
+
+		if fn == nil {
+			return
+		}
+		fn(ctx, d.throughput)
+
+		// Check to see if the context was canceled or
+		// the dispatcher was closed.
+		select {
+		case <-ctx.Done():
+			return
+		case <-d.closing:
+			return
+		default:
```

Comment:
is this to prevents the select statement from blocking? So when no communication can proceed it executes the default case and finishes immediately. But then the outer for loop will keep on running unless one of the select case is satisfied. If this is intended could you please put a log message/ comment in the default case?

Reply:
Yes, that is correct. Sure.